### PR TITLE
[CALCITE-5766] Add SAFE_NEGATE function (enabled for BigQuery library)

### PR DESCRIPTION
**Pull request by tanclary**
PR to contribute to https://issues.apache.org/jira/browse/CALCITE-5591

Any comments, questions, or suggestions appreciated as always!

[BigQuery docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions#safe_negate)

### CODE REVIEW

**Comment by tanclary** [comment]
@zoudan Here's SAFE_NEGATE, feel free to take a look!

**Comment by zoudan** [comment]
There are some checkstyle issues, could you please fix them?

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -1524,6 +1524,43 @@ public static int multiply(int b0, int b1) {
...
+  /** SQL <code>SAFE_NEGATE</code> function applied to a long value. */
+  public static @Nullable Long safeNegate(long b) {
+    BigDecimal ans = multiply(BigDecimal.valueOf(b), BigDecimal.valueOf(-1));
+    boolean overflow =
+        isOverflow(ans, BigDecimal.valueOf(Long.MAX_VALUE), BigDecimal.valueOf(Long.MIN_VALUE));
+    return overflow ? null : ans.longValue();
+  }
+
+  /** SQL <code>SAFE_NEGATE</code> function applied to a BigDecimal value. */
+  public static @Nullable BigDecimal safeNegate(BigDecimal b) {
```
Comment (zoudan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1557]: Should we add methods to handle other data types such as `float`, `smallint`?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -5243,6 +5243,23 @@ private static void checkIf(SqlOperatorFixture f) {
...
+    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.BIG_QUERY);
+    f.checkScalar("safe_negate(3)", "-3", "INTEGER");
+    f.checkScalar("safe_negate(-3)", "3", "INTEGER");
+    f.checkScalar("safe_negate(0)", "0", "INTEGER");
+    f.checkScalar("safe_negate(cast(3.0 as double))", "-3.0", "DOUBLE");
+    f.checkScalar("safe_negate(cast(3.456 as float))", "-3.456", "FLOAT");
+    f.checkScalar("safe_negate(cast(3.1234 as decimal(5, 4)))",
+        "-3.1234", "DECIMAL(5, 4)");
+    // check overflow returns null
+    f.checkNull("safe_negate(-9223372036854775808)");
```
Comment (zoudan) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5260]: Maybe we should add more overflow test cases for all kinds of numerical types？

**Review by julianhyde — COMMENTED** [review]
I think you need more tests - one for each numeric data type.

The implementation is actually straightforward:
For INTEGER, safe_negate(x) is `x == INTEGER.MIN_VALUE ? null : -x`

You probably don't need to add functions to `SqlFunctions`, just add logic in the implementor.

Or you could just implement `SAFE_NEGATE(x)` as `SAFE_MULTIPLY(x, -1)`.

**Comment by tanclary** [comment]
> I think you need more tests - one for each numeric data type.
> 
> The implementation is actually straightforward: For INTEGER, safe_negate(x) is `x == INTEGER.MIN_VALUE ? null : -x`
> 
> You probably don't need to add functions to `SqlFunctions`, just add logic in the implementor. Or you could just implement `SAFE_NEGATE(x)` as `SAFE_MULTIPLY(x, -1)`.
> 
> The key thing is a comprehensive set of tests.

@julianhyde I revisited this PR and made some of the changes you suggested. As far as tests go, I am not sure how to add tests for non-integer types given that the absolute value of the minimum and maximum values for other types are equivalent, so overflow, to my knowledge, is not possible. Let me know if you feel otherwise or have other comments.

**Comment by tanclary** [comment]
@julianhyde I rebased to include your recent changes and no longer use strings to identify the safe functions' methods. Let me know what you think.

**Comment by julianhyde** [comment]
Two cosmetic things in  testSafeNegateFunc:
 * There are too many '.' in the javadoc. The jira summary does not include a '.'
 * Function arguments should be indented 4, not aligned with the '('

+1 to merge without further review when you've fixed those

**Pull request merged**